Add SiStripCalCosmics ALCANANO

pieterdavid opened the pull request:
#### PR description:

This PR adds NanoAOD-style output for the SiStripCalCosmics ALCARECO (introduced in https://github.com/cms-sw/cmssw/pull/26032), based on the existing configuration (which was run by the DPG) to make flat trees for the Lorentz angle and backplane correction measurement, and taking advantage of the changes in https://github.com/cms-sw/cmssw/pull/33565 .
The code adds a plugin for per-run per-module information about the magnetic field etc., uses a `SimpleFlatTableProducer` for per-track information, and adds a plugin that stores the necessary per-cluster information (split in a base class that gets the clusters, and a subclass that stores the variables for the Lorentz angle and backplane correction - another subclass will be used for the gains tree).
This is a very small ALCARECO, so the size is small in both cases (the flat format is about the same size as the EDM format), but it has been extensively validated, and provides an interesting case to discuss the configuration and possibility of central production.
As a minimal solution to add this to cmsDriver I added a `SiStripCalCosmicsNano` ALCARECO based on the existing `SiStripCalCosmics`, and changed the code to insert a `NanoAODOutputModule` if the dataTier of a FilteredStream is NANOAOD. That seems to work (it can be tested with the `ALCA:SiStripCalCosmicsNano` step, which is also added to the relevant matrix workflows by this PR), but something less ad-hoc may be preferred - please provide input, this is the reason why the PR is marked as "work in progress".
The idea (discussed with XPOG) is to have this integrated for testing at T0 with CRAFT in October, so depending on if that works out and which release is used there, a backport to the 12_0 branch may be needed.

#### PR validation:

Event-by-event comparison for 2018 cosmics by @robervalwalsh , see [these slides](https://indico.cern.ch/event/992968/contributions/4424055/attachments/2270726/3856338/20210624_NanoAOD_CalibTrees.pdf)

CC: @robervalwalsh @mdelcourt @ataliercio @jandrejk @mmusich @tsusa @mariadalfonso @gouskos

pieterdavid committed:
Plugins and test config for NanoAOD-based calibtree from SiStripCalCosmics ALCARECO

pieterdavid committed:
Add SiStripCalCosmicsNano definition

pieterdavid committed:
Use NanoAODOutputModule for ALCARECO stream if dataTier equals NANOAOD

pieterdavid committed:
Add SiStripCalCosmicsNano to relval steps

pieterdavid committed:
Remove B-field filter

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34985/24829

- This PR adds an extra 100KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/PyReleaseValidation/python/relval_steps.py  modified in PR(s): #34714
  - File  Configuration/StandardSequences/python/AlCaRecoStreams_cff.py  modified in PR(s): #34965

cmsbuild commented:
A new Pull Request was created by  @pieterdavid (Pieter David) for master.

It involves the following packages:

- CalibTracker/SiStripCommon (**alca**)
- Calibration/TkAlCaRecoProducers (**alca**)
- Configuration/Applications (**operations**)
- Configuration/EventContent (**operations**)
- Configuration/PyReleaseValidation (**pdmv, upgrade**)
- Configuration/StandardSequences (**operations**)


@perrotta, @malbouis, @yuanchao, @jordan-martins, @davidlange6, @chayanit, @bbilin, @wajidalikhan, @kpedro88, @cmsbuild, @adrianodee, @srimanob, @kskovpen, @tlampen, @qliphy, @pohsun, @francescobrivio, @fabiocos, @tvami can you please review it and eventually sign? Thanks.
@echabert, @fabiocos, @gbenelli, @makortel, @felicepantaleo, @GiacomoSguazzoni, @JanFSchulte, @rovere, @VinInn, @Martin-Grunewald, @ebrondol, @tocheng, @lecriste, @mtosi, @mmusich, @threus, @dgulhan, @slomeo, @robervalwalsh this is something you requested to watch as well.
@perrotta, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

mmusich commented:
test parameters:
   * workflows = 7.2, 7.21, 7.22, 7.23, 7.24, 7.3, 7.4, 134.813, 136.733, 138.1 and 138.2

mmusich commented:
@cmsbuild, please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7e48dd/17976/summary.html
**COMMIT**: b1d1f0ba9746eed9e69111d75faf64fb8421b041
**CMSSW**: CMSSW_12_1_X_2021-08-23-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/34985/17976/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/134.813_RunCosmics2015C+RunCosmics2015C+RECOCOSDRUN2+ALCACOSDRUN2+HARVESTDCRUN2
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/136.733_RunCosmics2016B+RunCosmics2016B+RECOCOSDRUN2+ALCACOSDRUN2+HARVESTDCRUN2
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/138.1_RunCosmics2021+RunCosmics2021+RECOCOSDPROMPTRUN3+ALCACOSDRUN3+HARVESTDCRUN3
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/138.2_RunCosmics2021+RunCosmics2021+RECOCOSDEXPRUN3+ALCACOSDEXPRUN3+HARVESTDCEXPRUN3
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/7.21_Cosmics_UP17+Cosmics_UP17+DIGICOS_UP17+RECOCOS_UP17+ALCACOS_UP17+HARVESTCOS_UP17
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/7.22_Cosmics_UP16+Cosmics_UP16+DIGICOS_UP16+RECOCOS_UP16+ALCACOS_UP16+HARVESTCOS_UP16
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/7.23_Cosmics_UP21+Cosmics_UP21+DIGICOS_UP21+RECOCOS_UP21+ALCACOS_UP21+HARVESTCOS_UP21
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/7.24_Cosmics_UP21_0T+Cosmics_UP21_0T+DIGICOS_UP21_0T+RECOCOS_UP21_0T+ALCACOS_UP21_0T+HARVESTCOS_UP21_0T

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 39
 - DQMHistoTests: Total histograms compared: 3000352
 - DQMHistoTests: Total failures: 0
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3000330
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 38 files compared)
 - Checked 165 log files, 37 edm output root files, 39 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_1_X_2021-08-23-1100+7e48dd/44945/triggerResults): no differences found

pieterdavid committed:
Add SiStripCalCosmicsNano tests for Configuration/DataProcessing

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34985/24864

- This PR adds an extra 104KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/PyReleaseValidation/python/relval_steps.py  modified in PR(s): #34714

cmsbuild commented:
Pull request #34985 was updated. @perrotta, @malbouis, @yuanchao, @jordan-martins, @davidlange6, @chayanit, @bbilin, @wajidalikhan, @kpedro88, @cmsbuild, @adrianodee, @srimanob, @kskovpen, @tlampen, @qliphy, @pohsun, @francescobrivio, @fabiocos, @tvami can you please check and sign again.

mmusich commented:
@cmsbuild, please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7e48dd/18022/summary.html
**COMMIT**: 57212d5784deeea769bd98b082e7bfb1ed91c38e
**CMSSW**: CMSSW_12_1_X_2021-08-24-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/34985/18022/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/134.813_RunCosmics2015C+RunCosmics2015C+RECOCOSDRUN2+ALCACOSDRUN2+HARVESTDCRUN2
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/136.733_RunCosmics2016B+RunCosmics2016B+RECOCOSDRUN2+ALCACOSDRUN2+HARVESTDCRUN2
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/138.1_RunCosmics2021+RunCosmics2021+RECOCOSDPROMPTRUN3+ALCACOSDRUN3+HARVESTDCRUN3
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/138.2_RunCosmics2021+RunCosmics2021+RECOCOSDEXPRUN3+ALCACOSDEXPRUN3+HARVESTDCEXPRUN3
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/7.21_Cosmics_UP17+Cosmics_UP17+DIGICOS_UP17+RECOCOS_UP17+ALCACOS_UP17+HARVESTCOS_UP17
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/7.22_Cosmics_UP16+Cosmics_UP16+DIGICOS_UP16+RECOCOS_UP16+ALCACOS_UP16+HARVESTCOS_UP16
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/7.23_Cosmics_UP21+Cosmics_UP21+DIGICOS_UP21+RECOCOS_UP21+ALCACOS_UP21+HARVESTCOS_UP21
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/7.24_Cosmics_UP21_0T+Cosmics_UP21_0T+DIGICOS_UP21_0T+RECOCOS_UP21_0T+ALCACOS_UP21_0T+HARVESTCOS_UP21_0T

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 39
 - DQMHistoTests: Total histograms compared: 3000352
 - DQMHistoTests: Total failures: 0
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3000330
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 38 files compared)
 - Checked 165 log files, 37 edm output root files, 39 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_1_X_2021-08-24-1100+7e48dd/44986/triggerResults): no differences found

mmusich commented:
@pieterdavid this PR has now conflicts to solve.

tvami commented:
+alca

tvami commented on CalibTracker/SiStripCommon/test/MakeCalibrationTrees/produceCalibrationTree_CosmicsLABP_cfg.py:
```diff
@@ -0,0 +1,173 @@
+
+process.source = cms.Source("PoolSource", fileNames=cms.untracked.vstring(options.inputFiles))
+if options.runNumber != -1:
+   if 'Cosmics' not in options.inputCollection:  # FIXME: this should be improved
```
@pieterdavid

pieterdavid replied:
Sorry for the delay, yes, it's fine like this (I added the comment when I didn't understand these lines and thought more options were needed); it's a config for testing, so if you prefer I can also remove it

pieterdavid committed:
Remove a comment

pieterdavid committed:
update inspectNanoFile script for Runs and LuminosityBlocks trees

They are only added to the outputs if they contain non-standard products
(i.e. more branches than the run and lumiblock number)

pieterdavid commented:
Sorry for taking some time to update; I rebased on yesterday's IB and addressed a few comments raised by email. There is also a discussion on this planned in this afternoon's XPOG meeting https://indico.cern.ch/event/1066775/#3-sistrip-alcanano

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34985/24992

- This PR adds an extra 88KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/EventContent/python/AlCaRecoOutput_cff.py  modified in PR(s): #35029
  - File  Configuration/PyReleaseValidation/python/relval_steps.py  modified in PR(s): #35028
  - File  Configuration/StandardSequences/python/AlCaRecoStreams_cff.py  modified in PR(s): #35029

cmsbuild commented:
Pull request #34985 was updated. @gouskos, @chayanit, @bbilin, @wajidalikhan, @kpedro88, @tlampen, @pohsun, @perrotta, @yuanchao, @cmsbuild, @davidlange6, @AdrianoDee, @kskovpen, @qliphy, @fabiocos, @francescobrivio, @malbouis, @jordan-martins, @srimanob, @fgolf, @mariadalfonso, @tvami can you please check and sign again.

srimanob commented:
@cmsbuild please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7e48dd/18203/summary.html
**COMMIT**: 70e5bf9bc7df75a1692ed2a57799c2698f05740e
**CMSSW**: CMSSW_12_1_X_2021-08-31-2300/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/34985/18203/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/134.813_RunCosmics2015C+RunCosmics2015C+RECOCOSDRUN2+ALCACOSDRUN2+HARVESTDCRUN2
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/136.733_RunCosmics2016B+RunCosmics2016B+RECOCOSDRUN2+ALCACOSDRUN2+HARVESTDCRUN2
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/138.1_RunCosmics2021+RunCosmics2021+RECOCOSDPROMPTRUN3+ALCACOSDRUN3+HARVESTDCRUN3
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/138.2_RunCosmics2021+RunCosmics2021+RECOCOSDEXPRUN3+ALCACOSDEXPRUN3+HARVESTDCEXPRUN3
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/7.21_Cosmics_UP17+Cosmics_UP17+DIGICOS_UP17+RECOCOS_UP17+ALCACOS_UP17+HARVESTCOS_UP17
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/7.22_Cosmics_UP16+Cosmics_UP16+DIGICOS_UP16+RECOCOS_UP16+ALCACOS_UP16+HARVESTCOS_UP16
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/7.23_Cosmics_UP21+Cosmics_UP21+DIGICOS_UP21+RECOCOS_UP21+ALCACOS_UP21+HARVESTCOS_UP21
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/7.24_Cosmics_UP21_0T+Cosmics_UP21_0T+DIGICOS_UP21_0T+RECOCOS_UP21_0T+ALCACOS_UP21_0T+HARVESTCOS_UP21_0T

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 7 differences found in the comparisons
 - DQMHistoTests: Total files compared: 39
 - DQMHistoTests: Total histograms compared: 3000404
 - DQMHistoTests: Total failures: 11
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 3000370
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.004 KiB( 38 files compared)
 - DQMHistoSizes: changed ( 312.0 ): 0.004 KiB        MessageLogger/Warnings
 - Checked 165 log files, 37 edm output root files, 39 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_1_X_2021-08-31-2300+7e48dd/45137/triggerResults): no differences found

tvami commented:
+alca

mariadalfonso commented:
+xpog

changes in PhysicsTools/NanoAOD/test/inspectNanoFile.py
see result here (the lumi block is now monitored also in the standard nano)
https://cms-nanoaod-integration.web.cern.ch/integration/34985/data106Xul17v2_size_report.html

srimanob commented:
+Upgrade

mmusich commented:
@pieterdavid looks like a new rebase is needed.

qliphy commented:
@pieterdavid  Please resolve the conflict.

pieterdavid commented:
Thanks @mmusich and @qliphy , rebased (the conflict was a minor one with https://github.com/cms-sw/cmssw/pull/35202 in the relval steps definitions)

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34985/25245

- This PR adds an extra 108KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/PyReleaseValidation/python/relval_steps.py  modified in PR(s): #35028

cmsbuild commented:
Pull request #34985 was updated. @perrotta, @malbouis, @gouskos, @jordan-martins, @davidlange6, @chayanit, @bbilin, @wajidalikhan, @cmsbuild, @AdrianoDee, @srimanob, @yuanchao, @fgolf, @kskovpen, @qliphy, @mariadalfonso, @francescobrivio, @fabiocos, @tvami can you please check and sign again.

mmusich commented:
@cmsbuild, please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7e48dd/18580/summary.html
**COMMIT**: c52defb96dd625233f218921184017b1fed4e195
**CMSSW**: CMSSW_12_1_X_2021-09-13-2300/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/34985/18580/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/134.813_RunCosmics2015C+RunCosmics2015C+RECOCOSDRUN2+ALCACOSDRUN2+HARVESTDCRUN2
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/136.733_RunCosmics2016B+RunCosmics2016B+RECOCOSDRUN2+ALCACOSDRUN2+HARVESTDCRUN2
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/138.1_RunCosmics2021+RunCosmics2021+RECOCOSDPROMPTRUN3+ALCACOSDPROMPTRUN3+HARVESTDCPROMPTRUN3
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/138.2_RunCosmics2021+RunCosmics2021+RECOCOSDEXPRUN3+ALCACOSDEXPRUN3+HARVESTDCEXPRUN3
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/7.21_Cosmics_UP17+Cosmics_UP17+DIGICOS_UP17+RECOCOS_UP17+ALCACOS_UP17+HARVESTCOS_UP17
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/7.22_Cosmics_UP16+Cosmics_UP16+DIGICOS_UP16+RECOCOS_UP16+ALCACOS_UP16+HARVESTCOS_UP16
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/7.23_Cosmics_UP21+Cosmics_UP21+DIGICOS_UP21+RECOCOS_UP21+ALCACOS_UP21+HARVESTCOS_UP21
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-7e48dd/7.24_Cosmics_UP21_0T+Cosmics_UP21_0T+DIGICOS_UP21_0T+RECOCOS_UP21_0T+ALCACOS_UP21_0T+HARVESTCOS_UP21_0T

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 39
 - DQMHistoTests: Total histograms compared: 3001001
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3000973
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 38 files compared)
 - Checked 165 log files, 37 edm output root files, 39 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_1_X_2021-09-13-2300+7e48dd/45430/triggerResults): no differences found

tvami commented:
+alca

 * this was just rebased since the last signiture

tvami commented:
Hi @cms-sw/pdmv-l2 colleagues, what's your take of this PR? It's been around for a while now :)

kskovpen commented:
+1

srimanob commented:
+Upgrade

tvami commented:
> +xpog
> 
> changes in PhysicsTools/NanoAOD/test/inspectNanoFile.py
> see result here (the lumi block is now monitored also in the standard nano)
> https://cms-nanoaod-integration.web.cern.ch/integration/34985/data106Xul17v2_size_report.html

@mariadalfonso would you mind signing this again? The PR has been rebased since you did it already

mariadalfonso commented:
+xpog

qliphy commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will be automatically merged.

mmusich commented:
@perrotta @qliphy @tvami 
would it make sense to backport this to 12_0_X? 
CRAFT would be a perfect test bench.

tvami commented:
I agree that it would be nice to have this for CRAFT.
